cpp/src/rolling/rolling.cu

```diff
@@ -38,14 +39,127 @@ namespace experimental {
 namespace detail {
 
 namespace { // anonymous
+/**---------------------------------------------------------------------------*
+ * @brief Only count operation is executed and count is updated
+ *        depending on `min_periods` and returns true if it was
+ *        valid, else false.
+ *---------------------------------------------------------------------------**/
+template <typename InputType, typename OutputType, typename agg_op, aggregation::Kind op, bool has_nulls>
+std::enable_if_t<op == aggregation::COUNT, bool>
+__device__
+specific_rolling_kernel(column_device_view input,
+                        mutable_column_device_view output,
+                        size_type start_index,
+                        size_type end_index,
+                        size_type current_index,
+                        size_type min_periods,
+                        InputType identity) {
+
+    // declare this as volatile to avoid some compiler optimizations that lead to incorrect results
+    // for CUDA 10.0 and below (fixed in CUDA 10.1)
+    volatile cudf::size_type count = 0;
+    
+    for (size_type j = start_index; j < end_index; j++) {
+        if (!has_nulls || input.is_valid(j)) {
+            count++;
+        }
+    }
+   
+    bool output_is_valid = (count >= min_periods);
+    output.element<OutputType>(current_index) = count;
+
+    return output_is_valid;
+}
+
+/**---------------------------------------------------------------------------*
+ * @brief Only used for `string_view` type to get ARGMIN and ARGMAX, which
+ *        will be used to gather MIN and MAX. And returns true if the
+ *        operation was valid, else false.
+ *---------------------------------------------------------------------------**/
+template <typename InputType, typename OutputType, typename agg_op, aggregation::Kind op, bool has_nulls>
+std::enable_if_t<(op == aggregation::ARGMIN  or op == aggregation::ARGMAX) and
+                 std::is_same<InputType, cudf::string_view>::value, bool>
+__device__
+specific_rolling_kernel(column_device_view input,
+                        mutable_column_device_view output,
+                        size_type start_index,
+                        size_type end_index,
+                        size_type current_index,
+                        size_type min_periods,
+                        InputType identity) {
+
+    // declare this as volatile to avoid some compiler optimizations that lead to incorrect results
+    // for CUDA 10.0 and below (fixed in CUDA 10.1)
+    volatile cudf::size_type count = 0;
+    InputType val = identity;
+    OutputType val_index = (op == aggregation::ARGMIN)? ARGMIN_SENTINEL : ARGMAX_SENTINEL;
+
+    for (size_type j = start_index; j < end_index; j++) {
+        if (!has_nulls || input.is_valid(j)) {
+            InputType element = input.element<InputType>(j);
+            val = agg_op{}(element, val);
+            if (val == element) {
```

Comment:
Shouldn't the argmin/argmax index tracking be done inside the argmin/argmax aggregations?

Reply:
argmin/argmax aggregations updates the corresponding validity and value several times even though you might not need it for the rolling scenario. And this would be better than that.